File: asml/LC108ConvertSortedArrayToBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        if not nums:
            return None
        if len(nums) == 1:
            return TreeNode(nums[0])
        length = len(nums)
        root = TreeNode(nums[length/2])
        root.left = self.sortedArrayToBST(nums[:length/2])
        root.right = self.sortedArrayToBST(nums[length/2+1:])
        return root

File: asml/LC108ConvertSortedArrayToBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) return NULL;
        if (nums.size() == 1) return new TreeNode(nums[0]);
        int middle = nums.size()/2;
        TreeNode* root = new TreeNode(nums[middle]);
        vector<int> leftside(nums.begin(), nums.begin()+middle);
        vector<int> rightside(nums.begin()+middle+1, nums.end());
        root->left = sortedArrayToBST(leftside);
        root->right = sortedArrayToBST(rightside);
        return root;
    }
};
